<a href="https://colab.research.google.com/github/JainAnki/SiameseNetworks/blob/main/5a.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Installing pandas-profiling library

In [1]:
! pip install https://github.com/pandas-profiling/pandas-profiling/archive/master.zip

  Using cached https://github.com/pandas-profiling/pandas-profiling/archive/master.zip


#Importing Libraries

In [2]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import pandas_profiling as pp
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder

In [3]:
#%matplotlib inline

import matplotlib
# Force matplotlib to not use any Xwindows backend.
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(color_codes=True)
from scipy import stats

#import plotly.plotly as py

import plotly.graph_objs as go
from plotly.offline import init_notebook_mode,iplot
#init_notebook_mode(connected=True)

#Reading the "loan_train.csv" file
###1.   Read .csv file
###2.   Print the info of the dataframe
###3.   Print the top 5 rows of the dataframe



In [4]:
df = pd.read_csv('/content/loan_train.csv', low_memory=False)
df = df[(df['loan_status'] == 'Fully Paid') | (df['loan_status'] == 'Charged Off')]
print("df.info()")
print(df.info())
print("\n")
print("df.head()")
print(df.head())
print("\n")
print(df.shape)

df.info()
<class 'pandas.core.frame.DataFrame'>
Int64Index: 24301 entries, 0 to 24998
Columns: 111 entries, id to total_il_high_credit_limit
dtypes: float64(74), int64(13), object(24)
memory usage: 20.8+ MB
None


df.head()
       id  member_id  ...  total_bc_limit  total_il_high_credit_limit
0  573354     737474  ...             NaN                         NaN
1  476321     603324  ...             NaN                         NaN
2  451484     556265  ...             NaN                         NaN
4  800018    1005270  ...             NaN                         NaN
5  471391     595223  ...             NaN                         NaN

[5 rows x 111 columns]


(24301, 111)


#Creating and saving the ProfileReport of the **original** dataframe 



In [5]:
'''
profile = pp.ProfileReport(df, title = 'Loan_dataset', html = {'style': {'full_width': True}})
profile.to_notebook_iframe()
profile.to_file(output_file = 'Loan_dataset.html')
'''

"\nprofile = pp.ProfileReport(df, title = 'Loan_dataset', html = {'style': {'full_width': True}})\nprofile.to_notebook_iframe()\nprofile.to_file(output_file = 'Loan_dataset.html')\n"

#Reading the "loan_train.csv" file
###1.   Read .csv file to perform exploratory data analysis (EDA)
###2.   From the ProfileReport it was seen that a number of attributes had


*   only NaN values
*   a single value for all the data observations
  ### So we remove these attributes from the dataframe




In [6]:
df_re = pd.read_csv('/content/loan_train.csv', low_memory=False)
df_re = df_re[(df_re['loan_status'] == 'Fully Paid') | (df_re['loan_status'] == 'Charged Off')]
mapping_dict = {'loan_status':{'Fully Paid':1,'Charged Off': 0}}
df_re = df_re.replace(mapping_dict)
df_re.loan_status.value_counts()
df_re = df_re.drop(['next_pymnt_d','id','member_id','pymnt_plan','url','initial_list_status','mths_since_last_major_derog','policy_code','application_type','annual_inc_joint','dti_joint','verification_status_joint','acc_now_delinq','tot_coll_amt','tot_cur_bal','open_acc_6m','open_il_6m','open_il_12m', 'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il','il_util',	'open_rv_12m','open_rv_24m','max_bal_bc','all_util','total_rev_hi_lim', 'inq_fi','total_cu_tl','inq_last_12m','acc_open_past_24mths','avg_cur_bal','bc_open_to_buy','bc_util','delinq_amnt','mo_sin_old_il_acct','mo_sin_old_rev_tl_op','mo_sin_rcnt_rev_tl_op','mo_sin_rcnt_tl','mort_acc','mths_since_recent_bc','mths_since_recent_bc_dlq', 'mths_since_recent_inq', 'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd','num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats','num_bc_tl','num_il_tl','num_op_rev_tl','num_rev_accts','num_rev_tl_bal_gt_0',		'num_sats',	'num_tl_120dpd_2m','num_tl_30dpd','num_tl_90g_dpd_24m',	'num_tl_op_past_12m',	'pct_tl_nvr_dlq','percent_bc_gt_75','tot_hi_cred_lim','total_bal_ex_mort','total_bc_limit',	'total_il_high_credit_limit'], axis =1)

df_re_te = pd.read_csv('/content/loan_test.csv', low_memory=False)
df_re_te = df_re_te[(df_re_te['loan_status'] == 'Fully Paid') | (df_re_te['loan_status'] == 'Charged Off')]
mapping_dict = {'loan_status':{'Fully Paid':1,'Charged Off': 0}}
df_re_te = df_re_te.replace(mapping_dict)
df_re_te.loan_status.value_counts()
df_re_te = df_re_te.drop(['next_pymnt_d','id','member_id','pymnt_plan','url','initial_list_status','mths_since_last_major_derog','policy_code','application_type','annual_inc_joint','dti_joint','verification_status_joint','acc_now_delinq','tot_coll_amt','tot_cur_bal','open_acc_6m','open_il_6m','open_il_12m', 'open_il_24m', 'mths_since_rcnt_il', 'total_bal_il','il_util',	'open_rv_12m','open_rv_24m','max_bal_bc','all_util','total_rev_hi_lim', 'inq_fi','total_cu_tl','inq_last_12m','acc_open_past_24mths','avg_cur_bal','bc_open_to_buy','bc_util','delinq_amnt','mo_sin_old_il_acct','mo_sin_old_rev_tl_op','mo_sin_rcnt_rev_tl_op','mo_sin_rcnt_tl','mort_acc','mths_since_recent_bc','mths_since_recent_bc_dlq', 'mths_since_recent_inq', 'mths_since_recent_revol_delinq', 'num_accts_ever_120_pd','num_actv_bc_tl', 'num_actv_rev_tl', 'num_bc_sats','num_bc_tl','num_il_tl','num_op_rev_tl','num_rev_accts','num_rev_tl_bal_gt_0',		'num_sats',	'num_tl_120dpd_2m','num_tl_30dpd','num_tl_90g_dpd_24m',	'num_tl_op_past_12m',	'pct_tl_nvr_dlq','percent_bc_gt_75','tot_hi_cred_lim','total_bal_ex_mort','total_bc_limit',	'total_il_high_credit_limit'], axis =1)



##From the ProfileReport it was also seen that some of the attributes have
*   NaN + missing values as can be seen below

So these attributes were also removed and the number of attributes have now reduced from 111 to 45.





In [7]:
print('df_re[collections_12_mths_ex_med].unique()')
print(df_re['collections_12_mths_ex_med'].unique())
print('\n')
print('df_re[chargeoff_within_12_mths].unique()')
print(df_re['chargeoff_within_12_mths'].unique())
print('\n')
print('df_re[tax_liens].unique()')
print(df_re['tax_liens'].unique())
df_re = df_re.drop(['collections_12_mths_ex_med','chargeoff_within_12_mths','tax_liens'], axis= 1)
print('\n')
print('df_re.shape = ', df_re.shape)

print('df_re_te[collections_12_mths_ex_med].unique()')
print(df_re_te['collections_12_mths_ex_med'].unique())
print('\n')
print('df_re_te[chargeoff_within_12_mths].unique()')
print(df_re_te['chargeoff_within_12_mths'].unique())
print('\n')
print('df_re_te[tax_liens].unique()')
print(df_re_te['tax_liens'].unique())
df_re_te = df_re_te.drop(['collections_12_mths_ex_med','chargeoff_within_12_mths','tax_liens'], axis= 1)
print('\n')
print('df_re_te.shape = ', df_re_te.shape)

df_re[collections_12_mths_ex_med].unique()
[ 0. nan]


df_re[chargeoff_within_12_mths].unique()
[ 0. nan]


df_re[tax_liens].unique()
[ 0. nan]


df_re.shape =  (24301, 44)
df_re_te[collections_12_mths_ex_med].unique()
[ 0. nan]


df_re_te[chargeoff_within_12_mths].unique()
[ 0. nan]


df_re_te[tax_liens].unique()
[ 0. nan]


df_re_te.shape =  (14276, 44)


#Creating and saving the ProfileReport of the **reduced** dataframe 


In [8]:
'''
profile_re = pp.ProfileReport(df_re, title = 'Loan_dataset', html = {'style': {'full_width': True}})
profile_re.to_notebook_iframe()
profile_re.to_file(output_file = 'Loan_dataset_re.html')
'''

"\nprofile_re = pp.ProfileReport(df_re, title = 'Loan_dataset', html = {'style': {'full_width': True}})\nprofile_re.to_notebook_iframe()\nprofile_re.to_file(output_file = 'Loan_dataset_re.html')\n"

In [9]:
df_re.to_csv('cleaned_dataset.csv',index=False)

In [10]:
df_re.shape

(24301, 44)

#Using python to reduce the dataset instead of analyzing ProfileReport

In [11]:
df_co = pd.read_csv('/content/loan_train.csv', low_memory=False)
df_co_te = pd.read_csv('/content/loan_test.csv', low_memory=False)

In [12]:
print(df_co.isnull().sum())
df_co = df_co[(df_co['loan_status'] == 'Fully Paid') | (df_co['loan_status'] == 'Charged Off')]
mapping_dict = {'loan_status':{'Fully Paid':1,'Charged Off': 0}}
df_co = df_co.replace(mapping_dict)
#df_co.loan_status.value_counts()

print(df_co_te.isnull().sum())
df_co_te = df_co_te[(df_co_te['loan_status'] == 'Fully Paid') | (df_co_te['loan_status'] == 'Charged Off')]
mapping_dict_te = {'loan_status':{'Fully Paid':1,'Charged Off': 0}}
df_co_te = df_co_te.replace(mapping_dict_te)


id                                0
member_id                         0
loan_amnt                         0
funded_amnt                       0
funded_amnt_inv                   0
                              ...  
tax_liens                        17
tot_hi_cred_lim               24999
total_bal_ex_mort             24999
total_bc_limit                24999
total_il_high_credit_limit    24999
Length: 111, dtype: int64
id                                0
member_id                         0
loan_amnt                         0
funded_amnt                       0
funded_amnt_inv                   0
                              ...  
tax_liens                        22
tot_hi_cred_lim               14718
total_bal_ex_mort             14718
total_bc_limit                14718
total_il_high_credit_limit    14718
Length: 111, dtype: int64


In [13]:
df_co.shape
df_co_te.shape

(14276, 111)

In [14]:

drop_columns = []
for col in df_co.columns:
    non_null = df_co[col].dropna()
    unique_non_null = non_null.unique()
    num_true_unique = len(unique_non_null)
    if num_true_unique < 2:
        drop_columns.append(col)
df_co = df_co.drop(drop_columns, axis=1)

drop_columns_te = []
for col in df_co_te.columns:
    non_null = df_co_te[col].dropna()
    unique_non_null = non_null.unique()
    num_true_unique = len(unique_non_null)
    if num_true_unique < 2:
        drop_columns_te.append(col)
df_co_te = df_co_te.drop(drop_columns, axis=1)

In [15]:
df_co = df_co.drop([ 'id','member_id','url'], axis= 1)
df_co_te = df_co_te.drop([ 'id','member_id','url'], axis= 1)

In [16]:
half_count = len(df_co) / 2
df_co = df_co.dropna(thresh=half_count,axis=1)
 # Drop any column with more than 50% missing values

half_count = len(df_co_te) / 2
df_co_te = df_co_te.dropna(thresh=half_count,axis=1)
 # Drop any column with more than 50% missing values

In [17]:
df_co.shape
df_co_te.shape

(14276, 40)

In [18]:
df_co.select_dtypes(include=['object'])
df_co_te.select_dtypes(include=['object'])

term int_rate grade  ... revol_util last_pymnt_d last_credit_pull_d
0       36 months    6.62%     A  ...     11.60%       14-Nov             16-May
1       36 months   13.49%     C  ...     11.90%       14-May             14-Sep
2       36 months    8.94%     A  ...     46.40%       10-Feb             16-May
3       60 months   16.49%     D  ...     60.20%       16-May             16-May
4       36 months   15.21%     D  ...     55.70%       13-Sep             16-May
...           ...      ...   ...  ...        ...          ...                ...
14713   36 months    7.75%     A  ...     16.50%       10-Jul             13-Dec
14714   36 months    8.70%     B  ...      7.80%       10-Jul             10-Jul
14715   36 months    8.07%     A  ...     26.30%       10-Jul              7-Jun
14716   36 months    8.07%     A  ...     19.40%        8-Apr              7-Jun
14717   36 months    7.43%     A  ...      0.70%        8-Jan              7-Jun

[14276 rows x 18 columns]

#Transforming 'objects' with 'float' values to 'float' data type
##1. 'int_rate'
##2. 'revol_util'
##3. 'term'
##4. 'emp_length'

In [19]:
df_co['int_rate'].replace(to_replace='[^0-9.]', value='', inplace=True, regex=True)
df_co['int_rate'] = df_co['int_rate'].astype(float)

df_co['revol_util'].replace(to_replace='[^0-9.]', value='', inplace=True, regex=True)
df_co['revol_util'] = df_co['revol_util'].astype(float)

df_co['term'].replace(to_replace='[^0-9]+', value='', inplace=True, regex=True)
df_co['term'] = df_co['term'].astype(int)
set(df_co["term"])

#display(set(df_co["emp_length"]))
df_co.replace('n/a', np.nan,inplace=True)
df_co.emp_length.fillna(value=0,inplace=True)
set(df_co["emp_length"])
df_co['emp_length'].replace(to_replace='[^0-9]+', value='', inplace=True, regex=True)
df_co['emp_length'] = df_co['emp_length'].astype(int)
set(df_co["emp_length"])


df_co_te['int_rate'].replace(to_replace='[^0-9.]', value='', inplace=True, regex=True)
df_co_te['int_rate'] = df_co_te['int_rate'].astype(float)

df_co_te['revol_util'].replace(to_replace='[^0-9.]', value='', inplace=True, regex=True)
df_co_te['revol_util'] = df_co_te['revol_util'].astype(float)

df_co_te['term'].replace(to_replace='[^0-9]+', value='', inplace=True, regex=True)
df_co_te['term'] = df_co_te['term'].astype(int)
set(df_co_te["term"])

#display(set(df_co["emp_length"]))
df_co_te.replace('n/a', np.nan,inplace=True)
df_co_te.emp_length.fillna(value=0,inplace=True)
set(df_co_te["emp_length"])
df_co_te['emp_length'].replace(to_replace='[^0-9]+', value='', inplace=True, regex=True)
df_co_te['emp_length'] = df_co_te['emp_length'].astype(int)
set(df_co_te["emp_length"])

{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}

In [20]:
df_co.select_dtypes(include=['object'])
df_co_te.select_dtypes(include=['object'])

grade sub_grade  ... last_pymnt_d last_credit_pull_d
0         A        A2  ...       14-Nov             16-May
1         C        C2  ...       14-May             14-Sep
2         A        A5  ...       10-Feb             16-May
3         D        D3  ...       16-May             16-May
4         D        D2  ...       13-Sep             16-May
...     ...       ...  ...          ...                ...
14713     A        A3  ...       10-Jul             13-Dec
14714     B        B1  ...       10-Jul             10-Jul
14715     A        A4  ...       10-Jul              7-Jun
14716     A        A4  ...        8-Apr              7-Jun
14717     A        A2  ...        8-Jan              7-Jun

[14276 rows x 14 columns]

#Transforming Categorical values to 'str' data type
##1. grade
##2. sub_grade
##3. home_ownership
##4. verification_status

In [21]:
print(set(df_co['grade']),"\n",
set(df_co['sub_grade']),"\n",
set(df_co['home_ownership']),"\n",
set(df_co['verification_status']),"\n",
set(df_co['purpose']))
df_co['grade']= df_co['grade'].convert_dtypes()
df_co['sub_grade']=df_co['sub_grade'].convert_dtypes()
df_co['home_ownership'] = df_co['home_ownership'].convert_dtypes()
df_co['verification_status'] = df_co['verification_status'].convert_dtypes()
df_co['purpose']=df_co['purpose'].convert_dtypes()
df_co.select_dtypes(include=['object'])

print(set(df_co_te['grade']),"\n",
set(df_co_te['sub_grade']),"\n",
set(df_co_te['home_ownership']),"\n",
set(df_co_te['verification_status']),"\n",
set(df_co_te['purpose']))
df_co_te['grade']= df_co_te['grade'].convert_dtypes()
df_co_te['sub_grade']=df_co_te['sub_grade'].convert_dtypes()
df_co_te['home_ownership'] = df_co_te['home_ownership'].convert_dtypes()
df_co_te['verification_status'] = df_co_te['verification_status'].convert_dtypes()
df_co_te['purpose']=df_co['purpose'].convert_dtypes()
df_co_te.select_dtypes(include=['object'])

{'C', 'F', 'A', 'E', 'D', 'B', 'G'} 
 {'G3', 'G5', 'C3', 'A2', 'F3', 'A4', 'G4', 'A1', 'D4', 'B3', 'D5', 'F4', 'G2', 'B4', 'G1', 'E1', 'D2', 'C2', 'D3', 'D1', 'E4', 'F5', 'A5', 'F2', 'B2', 'B1', 'C5', 'E2', 'C4', 'A3', 'C1', 'E3', 'F1', 'E5', 'B5'} 
 {'MORTGAGE', 'OTHER', 'OWN', 'RENT'} 
 {'Source Verified', 'Verified', 'Not Verified'} 
 {'medical', 'renewable_energy', 'major_purchase', 'other', 'moving', 'credit_card', 'house', 'vacation', 'educational', 'wedding', 'small_business', 'car', 'home_improvement', 'debt_consolidation'}
{'C', 'F', 'A', 'E', 'D', 'B', 'G'} 
 {'G3', 'G5', 'C3', 'A2', 'F3', 'A4', 'G4', 'A1', 'D4', 'B3', 'B4', 'D5', 'F4', 'G2', 'G1', 'E1', 'D2', 'C2', 'D3', 'D1', 'E4', 'F5', 'A5', 'F2', 'B2', 'C5', 'B1', 'E2', 'C4', 'A3', 'C1', 'E3', 'F1', 'E5', 'B5'} 
 {'NONE', 'MORTGAGE', 'OWN', 'OTHER', 'RENT'} 
 {'Source Verified', 'Verified', 'Not Verified'} 
 {'medical', 'renewable_energy', 'major_purchase', 'other', 'moving', 'credit_card', 'house', 'vacation', 'educatio

emp_title issue_d  ... last_pymnt_d last_credit_pull_d
0         PPG Industries, Inc.  11-Nov  ...       14-Nov             16-May
1      PPI Technology Services  11-Aug  ...       14-May             14-Sep
2              PPL Corporation   9-Sep  ...       10-Feb             16-May
3                       ppl eu  11-May  ...       16-May             16-May
4                         PPMM  10-Sep  ...       13-Sep             16-May
...                        ...     ...  ...          ...                ...
14713                      NaN   7-Jul  ...       10-Jul             13-Dec
14714                      NaN   7-Jul  ...       10-Jul             10-Jul
14715                      NaN   7-Jul  ...       10-Jul              7-Jun
14716                      NaN   7-Jul  ...        8-Apr              7-Jun
14717                      NaN   7-Jul  ...        8-Jan              7-Jun

[14276 rows x 9 columns]

In [22]:
print(df_co.isnull().sum())
print(df_co_te.isnull().sum())

loan_amnt                     0
funded_amnt                   0
funded_amnt_inv               0
term                          0
int_rate                      0
installment                   0
grade                         0
sub_grade                     0
emp_title                     6
emp_length                    0
home_ownership                0
annual_inc                    0
verification_status           0
issue_d                       0
loan_status                   0
desc                       7943
purpose                       0
title                         8
zip_code                      0
addr_state                    0
dti                           0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                   29
total_acc                     0
total_pymnt                   0
total_pymnt_inv               0
total_re

In [23]:
df_co = df_co.fillna(df_co['emp_title'].value_counts().index[0])
for col in ['emp_title','issue_d','desc','title', 'zip_code','addr_state','earliest_cr_line','last_pymnt_d','revol_util','last_pymnt_amnt','last_credit_pull_d','pub_rec_bankruptcies']:
  df_co[col] = df_co[col].astype('category')
#labels = cat_df_flights['carrier'].astype('category').cat.categories.tolist()
#replace_map_comp = {'carrier' : {k: v for k,v in zip(labels,list(range(1,len(labels)+1)))}}
print(df_co.dtypes)
#print(replace_map_comp)

df_co_te = df_co_te.fillna(df_co_te['emp_title'].value_counts().index[0])
for col in ['emp_title','issue_d','desc','title', 'zip_code','addr_state','earliest_cr_line','last_pymnt_d','revol_util','last_pymnt_amnt','last_credit_pull_d','pub_rec_bankruptcies']:
  df_co_te[col] = df_co_te[col].astype('category')
#labels = cat_df_flights['carrier'].astype('category').cat.categories.tolist()
#replace_map_comp = {'carrier' : {k: v for k,v in zip(labels,list(range(1,len(labels)+1)))}}
print(df_co_te.dtypes)
#print(replace_map_comp)

loan_amnt                     int64
funded_amnt                   int64
funded_amnt_inv             float64
term                          int64
int_rate                    float64
installment                 float64
grade                        string
sub_grade                    string
emp_title                  category
emp_length                    int64
home_ownership               string
annual_inc                  float64
verification_status          string
issue_d                    category
loan_status                   int64
desc                       category
purpose                      string
title                      category
zip_code                   category
addr_state                 category
dti                         float64
delinq_2yrs                   int64
earliest_cr_line           category
inq_last_6mths                int64
open_acc                      int64
pub_rec                       int64
revol_bal                     int64
revol_util                 c

In [24]:
df_co1 = df_co.select_dtypes(include=['object']).copy()
for col in ['grade','sub_grade','emp_title','issue_d','desc','title', 'addr_state','earliest_cr_line','last_pymnt_d','revol_util','last_pymnt_amnt','last_credit_pull_d','pub_rec_bankruptcies']:
  le = LabelEncoder()
  #df_co['emp_title']=df_co['emp_title'].convert_dtypes()
  #df_co.dtypes
  df_co[col] = le.fit_transform(df_co[col].astype(str))

#df_co1['emp_title'] = le.fit_transform(df_co['emp_title'])
#df2.value_counts()

df_co_te1 = df_co_te.select_dtypes(include=['object']).copy()
for col in ['grade','sub_grade','emp_title','issue_d','desc','title', 'addr_state','earliest_cr_line','last_pymnt_d','revol_util','last_pymnt_amnt','last_credit_pull_d','pub_rec_bankruptcies']:
  le = LabelEncoder()
  #df_co['emp_title']=df_co['emp_title'].convert_dtypes()
  #df_co.dtypes
  df_co_te[col] = le.fit_transform(df_co_te[col].astype(str))

#df_co1['emp_title'] = le.fit_transform(df_co['emp_title'])
#df2.value_counts()


In [25]:
df_co.dtypes
df_co_te.dtypes

loan_amnt                     int64
funded_amnt                   int64
funded_amnt_inv             float64
term                          int64
int_rate                    float64
installment                 float64
grade                         int64
sub_grade                     int64
emp_title                     int64
emp_length                    int64
home_ownership               string
annual_inc                  float64
verification_status          string
issue_d                       int64
loan_status                   int64
desc                          int64
purpose                      string
title                         int64
zip_code                   category
addr_state                    int64
dti                         float64
delinq_2yrs                   int64
earliest_cr_line              int64
inq_last_6mths                int64
open_acc                      int64
pub_rec                       int64
revol_bal                     int64
revol_util                  

#One Hot Encoding for nominal attributes
##1. purpose
##3. home_ownership
##4. verification_status

In [26]:
nom_cols = ['home_ownership','verification_status','purpose']
df_du = pd.get_dummies(df_co[nom_cols])
df_co = pd.concat([df_co, df_du], axis=1)
df_co = df_co.drop(nom_cols, axis=1)
df_co.head()

nom_cols1 = ['home_ownership','verification_status','purpose']
df_du1 = pd.get_dummies(df_co_te[nom_cols1])
df_co_te = pd.concat([df_co_te, df_du1], axis=1)

df_co_te = df_co_te.drop(nom_cols1, axis=1)
df_co_te.head()

loan_amnt  funded_amnt  ...  last_credit_pull_d  pub_rec_bankruptcies
0       4200         4200  ...                  76                     0
1       1325         1325  ...                  59                     0
2      10000        10000  ...                  76                     0
3       6000         6000  ...                  76                     0
4      25000        25000  ...                  76                     0

[5 rows x 37 columns]

In [27]:
display(df_co.describe(),
df_co.dtypes.value_counts())
obj = df_co.columns[df_co.dtypes == 'object']
df_co[obj].isnull().sum()

display(df_co_te.describe(),
df_co_te.dtypes.value_counts())
obj = df_co_te.columns[df_co_te.dtypes == 'object']
df_co_te[obj].isnull().sum()

loan_amnt   funded_amnt  ...  last_credit_pull_d  pub_rec_bankruptcies
count  24301.000000  24301.000000  ...        24301.000000          24301.000000
mean   11108.384429  10836.688819  ...           56.448212              0.092712
std     7295.677660   7032.562475  ...           20.133192              0.432911
min      500.000000    500.000000  ...            0.000000              0.000000
25%     5500.000000   5500.000000  ...           42.000000              0.000000
50%    10000.000000   9600.000000  ...           62.000000              0.000000
75%    15000.000000  15000.000000  ...           76.000000              0.000000
max    35000.000000  35000.000000  ...          101.000000              3.000000

[8 rows x 36 columns]

int64       24
float64     12
category     1
dtype: int64

loan_amnt   funded_amnt  ...  last_credit_pull_d  pub_rec_bankruptcies
count  14276.000000  14276.000000  ...        14276.000000          14276.000000
mean   10942.578453  10694.469739  ...           55.881549              0.104021
std     7436.489565   7186.893181  ...           20.346079              0.459386
min      500.000000    500.000000  ...            0.000000              0.000000
25%     5000.000000   5000.000000  ...           41.000000              0.000000
50%     9450.000000   9000.000000  ...           60.000000              0.000000
75%    15000.000000  15000.000000  ...           75.000000              0.000000
max    35000.000000  35000.000000  ...          103.000000              3.000000

[8 rows x 36 columns]

int64       24
float64     12
category     1
dtype: int64

Series([], dtype: float64)

In [28]:
df_co.shape
df_co_te.shape

(14276, 37)

In [29]:
df_co['zip_code'].unique()
df_co['zip_code'] = df_co['zip_code'].str.rstrip('x')
df_co['zip_code'].head()


df_co_te['zip_code'].unique()
df_co_te['zip_code'] = df_co_te['zip_code'].str.rstrip('x')
df_co_te['zip_code'].head()


0    152
1    705
2    181
3    187
4    950
Name: zip_code, dtype: object

In [30]:
df_co['zip_code'] = df_co['zip_code'].astype(int)
df_co_te['zip_code'] = df_co_te['zip_code'].astype(int)

In [31]:
df_co.describe()
df_co_te.describe()

loan_amnt   funded_amnt  ...  last_credit_pull_d  pub_rec_bankruptcies
count  14276.000000  14276.000000  ...        14276.000000          14276.000000
mean   10942.578453  10694.469739  ...           55.881549              0.104021
std     7436.489565   7186.893181  ...           20.346079              0.459386
min      500.000000    500.000000  ...            0.000000              0.000000
25%     5000.000000   5000.000000  ...           41.000000              0.000000
50%     9450.000000   9000.000000  ...           60.000000              0.000000
75%    15000.000000  15000.000000  ...           75.000000              0.000000
max    35000.000000  35000.000000  ...          103.000000              3.000000

[8 rows x 37 columns]

In [32]:
def zip_formatted(row):
    if len(str(row)) == 1:
        row = str(row)+ '00'
    elif len(str(row)) == 2:
        row = str(row)+ '0'
    return int(row)

In [33]:
df_co['zip_code'] = df_co['zip_code'].apply(zip_formatted)
df_co_te['zip_code'] = df_co_te['zip_code'].apply(zip_formatted)

In [34]:
min_value = df_co['loan_amnt'].min()
max_value = df_co['loan_amnt'].max()
print(min_value)
print(max_value)
bins = np.linspace(min_value,max_value,5)
print(len(bins))
labels = ['0','1','2','3']
df_co['bins'] = pd.cut(df_co['loan_amnt'], bins=bins, labels=labels, include_lowest=True)
df_co['bins']

min_value = df_co_te['loan_amnt'].min()
max_value = df_co_te['loan_amnt'].max()
print(min_value)
print(max_value)
bins = np.linspace(min_value,max_value,5)
print(len(bins))
labels = ['0','1','2','3']
df_co_te['bins'] = pd.cut(df_co_te['loan_amnt'], bins=bins, labels=labels, include_lowest=True)
df_co_te['bins']

500
35000
5
500
35000
5


0        0
1        0
2        1
3        0
4        2
        ..
14713    0
14714    0
14715    0
14716    0
14717    0
Name: bins, Length: 14276, dtype: category
Categories (4, object): ['0' < '1' < '2' < '3']

In [35]:
import matplotlib.pyplot as plt
plt.hist(df_co['bins'], bins=5)
plt.hist(df_co_te['bins'], bins=5)

(array([4677., 7073.,    0., 2006.,  520.]),
 array([0. , 0.6, 1.2, 1.8, 2.4, 3. ]),
 <a list of 5 Patch objects>)

In [36]:
bad_indicator = df_co["loan_status"].isin(["1"])
bad_indicator.value_counts()

bad_indicator_te = df_co_te["loan_status"].isin(["1"])
bad_indicator_te.value_counts()


True     12123
False     2153
Name: loan_status, dtype: int64

In [37]:
from sklearn.ensemble import RandomForestClassifier
X_tr= df_co.copy()
X_te = df_co_te.copy()
clf = RandomForestClassifier(max_depth=3)
clf = clf.fit(X_tr, bad_indicator)

In [38]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, X_tr, bad_indicator, cv=5, scoring='accuracy')
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std()))

Accuracy: 0.99 (+/- 0.00)


In [39]:
importances = clf.feature_importances_
std = np.std([tree.feature_importances_ for tree in clf.estimators_],
             axis=0)
indices = np.argsort(importances)[::-1]

# Print the feature ranking
print("Feature ranking:")

names = X_tr.columns

for f in range(X_tr.shape[1]):
    print("%d. %s (%f)" % (f + 1, names[indices[f]], importances[indices[f]]))

# Plot the feature importances of the forest
plt.figure()
plt.title("Feature importances")
plt.bar(range(X_tr.shape[1]), importances[indices],
       color="r", yerr=std[indices], align="center")
plt.xticks(range(X_tr.shape[1]), indices)
plt.xlim([-1, X_tr.shape[1]])
plt.show()

Feature ranking:
1. loan_status (0.333465)
2. recoveries (0.234303)
3. collection_recovery_fee (0.195965)
4. total_rec_prncp (0.095509)
5. total_pymnt (0.034301)
6. installment (0.020152)
7. total_pymnt_inv (0.019209)
8. loan_amnt (0.011415)
9. total_rec_late_fee (0.011404)
10. funded_amnt_inv (0.009602)
11. grade (0.006024)
12. total_rec_int (0.005595)
13. funded_amnt (0.005037)
14. int_rate (0.004794)
15. last_credit_pull_d (0.003164)
16. last_pymnt_d (0.003097)
17. term (0.003046)
18. sub_grade (0.002107)
19. last_pymnt_amnt (0.001014)
20. inq_last_6mths (0.000359)
21. bins (0.000121)
22. title (0.000076)
23. annual_inc (0.000040)
24. issue_d (0.000035)
25. pub_rec_bankruptcies (0.000031)
26. revol_util (0.000030)
27. dti (0.000027)
28. revol_bal (0.000025)
29. zip_code (0.000017)
30. desc (0.000013)
31. pub_rec (0.000012)
32. emp_length (0.000009)
33. delinq_2yrs (0.000001)
34. total_acc (0.000000)
35. addr_state (0.000000)
36. open_acc (0.000000)
37. earliest_cr_line (0.000000)
38

In [40]:
#a = df_co.corrwith(df_co['loan_status'])
X_tr= df_co.copy()
X_te = df_co_te.copy()

In [41]:
a = df_co.corrwith(df_co['loan_status'])
a

loan_amnt                 -0.058606
funded_amnt               -0.056209
funded_amnt_inv           -0.038255
term                      -0.169273
int_rate                  -0.210081
installment               -0.026918
grade                     -0.199933
sub_grade                 -0.202964
emp_title                 -0.028804
emp_length                -0.011834
annual_inc                 0.040116
issue_d                   -0.008817
loan_status                1.000000
desc                      -0.017400
title                     -0.047286
zip_code                  -0.020531
addr_state                 0.021916
dti                       -0.046135
delinq_2yrs               -0.013968
earliest_cr_line           0.013776
inq_last_6mths            -0.075122
open_acc                   0.000698
pub_rec                   -0.050736
revol_bal                 -0.006231
revol_util                -0.086858
total_acc                  0.017341
total_pymnt                0.233951
total_pymnt_inv            0

In [42]:
X_tr=X_tr.drop(['loan_amnt','funded_amnt','funded_amnt_inv','term','int_rate','installment','grade','sub_grade'],axis =1)
X_tr = X_tr.drop(['emp_title','emp_length','issue_d','title','zip_code','addr_state','dti','delinq_2yrs'],axis = 1)
X_tr = X_tr.drop(['earliest_cr_line','inq_last_6mths','pub_rec','revol_bal','revol_util','last_pymnt_amnt','pub_rec_bankruptcies'],axis = 1)
X_tr= X_tr.drop(['total_rec_int','total_rec_late_fee','recoveries','collection_recovery_fee'],axis =1)



X_te=X_te.drop(['loan_amnt','funded_amnt','funded_amnt_inv','term','int_rate','installment','grade','sub_grade'],axis =1)
X_te = X_te.drop(['emp_title','emp_length','issue_d','title','zip_code','addr_state','dti','delinq_2yrs'],axis = 1)
X_te = X_te.drop(['earliest_cr_line','inq_last_6mths','pub_rec','revol_bal','revol_util','last_pymnt_amnt','pub_rec_bankruptcies'],axis = 1)
X_te= X_te.drop(['total_rec_int','total_rec_late_fee','recoveries','collection_recovery_fee'],axis =1)


In [43]:
X_tr.to_csv('filtered_loans_train.csv',index=False)
X_tr.head()

X_te.to_csv('filtered_loans_test.csv',index=False)
X_te.head()

annual_inc  loan_status  desc  ...  last_pymnt_d  last_credit_pull_d  bins
0     69000.0            1  9377  ...            57                  76     0
1     24000.0            0  4175  ...            56                  59     0
2     75000.0            1  8850  ...             3                  76     1
3     60000.0            1  9377  ...            76                  76     0
4    135000.0            1  9377  ...            47                  76     2

[5 rows x 11 columns]

In [44]:
X_tr.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 24301 entries, 0 to 24998
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   annual_inc          24301 non-null  float64 
 1   loan_status         24301 non-null  int64   
 2   desc                24301 non-null  int64   
 3   open_acc            24301 non-null  int64   
 4   total_acc           24301 non-null  int64   
 5   total_pymnt         24301 non-null  float64 
 6   total_pymnt_inv     24301 non-null  float64 
 7   total_rec_prncp     24301 non-null  float64 
 8   last_pymnt_d        24301 non-null  int64   
 9   last_credit_pull_d  24301 non-null  int64   
 10  bins                24301 non-null  category
dtypes: category(1), float64(4), int64(6)
memory usage: 2.7 MB


In [45]:
X_te.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14276 entries, 0 to 14717
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   annual_inc          14276 non-null  float64 
 1   loan_status         14276 non-null  int64   
 2   desc                14276 non-null  int64   
 3   open_acc            14276 non-null  int64   
 4   total_acc           14276 non-null  int64   
 5   total_pymnt         14276 non-null  float64 
 6   total_pymnt_inv     14276 non-null  float64 
 7   total_rec_prncp     14276 non-null  float64 
 8   last_pymnt_d        14276 non-null  int64   
 9   last_credit_pull_d  14276 non-null  int64   
 10  bins                14276 non-null  category
dtypes: category(1), float64(4), int64(6)
memory usage: 1.2 MB


In [46]:
X_tr['loan_status'].value_counts() / X_tr.shape[0]


1    0.857043
0    0.142957
Name: loan_status, dtype: float64

In [47]:
def compute_metrics(predictions):
    
    predictions = pd.Series(predictions)
    # False positives
    fp_filter = (predictions == 1) & (loans['loan_status'] == 0)
    fp = len(predictions[fp_filter])
    
    # True positives
    tp_filter = (predictions == 1) & (loans['loan_status'] == 1)
    tp = len(predictions[tp_filter])
    
    # False negatives
    fn_filter = (predictions == 0) & (loans['loan_status'] == 1)
    fn = len(predictions[fn_filter])
    
    # True negatives
    tn_filter = (predictions == 0)& (loans['loan_status'] == 0)
    tn = len(predictions[tn_filter])
    
    # Rates
    tpr = tp / (tp + fn)
    fpr = fp / (fp + tn)
    
    print(tpr * 100)
    print(fpr * 100)
    print()
    
    print('Other Mesurements:')
    
    
    sp = tn / (tn + fp)
    sty = tp / (tp + fn)
    G_mean = np.sqrt(sp * sty)
    prcn = tp / (tp + fp)
    acc = (tn + tp) / features.shape[0]
    method_dict = {'specificity': '{:.2f}'.format(sp*100), 
               'sensitivity': '{:.2f}'.format(sty*100), 
               'G-mean': '{:.2f}'.format(G_mean*100),
               'precision': '{:.2f}'.format(prcn*100), 
               'accuracy': '{:.2f}'.format(acc*100),
               'False Positive Rate': '{:.2f}'.format(fpr*100)
              }
    return pd.DataFrame(method_dict,index=['values'])

In [48]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_predict, KFold

scaler = StandardScaler()

In [49]:

x_cols = X_tr.columns
features_cols = x_cols.drop(['loan_status'])
features = X_tr[features_cols]
target = X_tr['loan_status']


In [50]:
def getEvaluationMatrix(predicDF):
    lablePrediction = predicDF.select( "loan_status", "prediction")
    lablePrediction.cache()
    totalCount = lablePrediction.count()
    correctCount = lablePrediction.filter(col("loan_status") == col("prediction")).count()
    wrongCount = lablePrediction.filter(~(col("loan_status") == col("prediction"))).count()
    trueP = lablePrediction.filter(col("loan_status") == 0.0).filter(col("loan_status") == col("prediction")).count()
    trueN = lablePrediction.filter(col("loan_status") == 1.0).filter(col("loan_status") == col("prediction")).count()
    falseN = lablePrediction.filter(col("loan_status") == 1.0).filter(~(col("loan_status") == col("prediction"))).count()
    falseP = lablePrediction.filter(col("loan_status") == 0.0).filter(~(col("loan_status") == col("prediction"))).count()

    ratioWrong = float(wrongCount) / float(totalCount) 
    ratioCorrect = float(correctCount)/ float(totalCount)

    print("totalCount   - ", totalCount)
    print("correctCount - ", correctCount)
    print("wrongCount   - ", wrongCount)
    print("trueP        - ", trueP)
    print("trueN        - ", trueN)
    print("falseN       - ", falseN)
    print("falseP       - ", falseP)
    print("ratioWrong   - ", ratioWrong)
    print("ratioCorrect - ", ratioCorrect)
    
    precision = ((float(trueP) / (float(trueP) + float(falseP))) * 100 )
    recall = ((float(trueP) / (float(trueP) + float(falseN))) * 100 )
    print("Accuracy     - ", (trueP + trueN) / totalCount)
    print("Precision    - ", precision)
    print("Recall       - ", recall)
    print("F-1 Score    - ", ((2* ( (precision*recall) / (precision + recall))) ))
    print("Sensitivity  - ", ((float(trueP) / (float(trueP) + float(falseN))) * 100 ))
    print("Specificity  - ", ((float(trueN) / (float(trueN) + float(falseP))) * 100 ))
    
    createROC(predictions)

In [51]:
def createROC(predictions):
    results = predictions.select(['probability', 'loan_status'])
 
    ## prepare score-label set
    results_collect = results.collect()
    results_list = [(float(i[0][0]), 1.0-float(i[1])) for i in results_collect]
    scoreAndLabels = spark.sparkContext.parallelize(results_list)
 
    bcMetrics = BinaryClassificationMetrics(scoreAndLabels)
    print("ROC score is - ", bcMetrics.areaUnderROC)
        
    fpr = dict()
    tpr = dict()
    roc_auc = dict()
 
    y_test = [i[1] for i in results_list]
    y_score = [i[0] for i in results_list]
 
    fpr, tpr, _ = roc_curve(y_test, y_score)
    roc_auc = auc(fpr, tpr)
 
    plt.figure()
    plt.plot(fpr, tpr, label='ROC curve (area = %0.2f)' % roc_auc)
    plt.plot([0, 1], [0, 1], 'k--')
    plt.xlim([0.0, 1.0])
    plt.ylim([0.0, 1.05])
    plt.xlabel('False Positive Rate')
    plt.ylabel('True Positive Rate')
    plt.title('Receiver operating characteristic example')
    plt.legend(loc="lower right")
    plt.show()

In [52]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.metrics import classification_report

In [53]:
x_cols = X_tr.columns
features_cols = x_cols.drop(['loan_status'])
y_tr = X_tr['loan_status']
X_tr = X_tr[features_cols]

x_cols = X_te.columns
features_cols = x_cols.drop(['loan_status'])
y_te = X_te['loan_status']
X_te = X_te[features_cols]

# define the model
model = GradientBoostingClassifier(learning_rate = 0.1)
model.get_params()
# define the evaluation method
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)
# evaluate the model on the dataset
n_scores = cross_val_score(model, X_tr, y_tr, scoring='accuracy', cv=cv, n_jobs=-1)
# report performance
print('Mean Accuracy: %.3f (%.3f)' % (np.mean(n_scores), np.std(n_scores)))

#gradient_booster = GradientBoostingClassifier(learning_rate=0.1)
print(model.get_params())

model.fit(X_tr,y_tr)
print(classification_report(y_te,model.predict(X_te)))

Mean Accuracy: 0.966 (0.003)
{'ccp_alpha': 0.0, 'criterion': 'friedman_mse', 'init': None, 'learning_rate': 0.1, 'loss': 'deviance', 'max_depth': 3, 'max_features': None, 'max_leaf_nodes': None, 'min_impurity_decrease': 0.0, 'min_impurity_split': None, 'min_samples_leaf': 1, 'min_samples_split': 2, 'min_weight_fraction_leaf': 0.0, 'n_estimators': 100, 'n_iter_no_change': None, 'presort': 'deprecated', 'random_state': None, 'subsample': 1.0, 'tol': 0.0001, 'validation_fraction': 0.1, 'verbose': 0, 'warm_start': False}
              precision    recall  f1-score   support

           0       0.98      0.78      0.86      2153
           1       0.96      1.00      0.98     12123

    accuracy                           0.96     14276
   macro avg       0.97      0.89      0.92     14276
weighted avg       0.96      0.96      0.96     14276



In [54]:
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.model_selection import GridSearchCV

In [55]:
grid = dict()
grid['n_estimators'] = [10, 50, 100, 500]
grid['learning_rate'] = [0.01, 0.1]
grid['subsample'] = [0.5, 0.9]
scores = ["precision", "recall"]
# define the model with default hyperparameters
model = GradientBoostingClassifier()
# define the evaluation procedure
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

# define the grid search procedure
for score in scores:
    print("# Tuning hyper-parameters for %s" % score)
    print()
    grid_search = GridSearchCV(estimator=model, param_grid=grid, n_jobs=-1, cv=cv, scoring="%s_macro" % score)
    # execute the grid search
    grid_result = grid_search.fit(X_tr, y_tr)
    # summarize the best score and configuration
    print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
    # summarize all scores that were evaluated
    print("Grid scores on development set:")
    print()
    means = grid_result.cv_results_['mean_test_score']
    stds = grid_result.cv_results_['std_test_score']
    params = grid_result.cv_results_['params']
    for mean, stdev, param in zip(means, stds, params):
      print("%f (%f) with: %r" % (mean, stdev, param))
    print()

    print("Detailed classification report:")
    print()
    print("The model is trained on the full development set.")
    print("The scores are computed on the full evaluation set.")
    print()
    y_true, y_pred = y_te, grid_search.predict(X_te)
    print(classification_report(y_true, y_pred))
    print()

# Tuning hyper-parameters for precision

Best: 0.973535 using {'learning_rate': 0.1, 'n_estimators': 500, 'subsample': 0.9}
Grid scores on development set:

0.428521 (0.000098) with: {'learning_rate': 0.01, 'n_estimators': 10, 'subsample': 0.5}
0.428521 (0.000098) with: {'learning_rate': 0.01, 'n_estimators': 10, 'subsample': 0.9}
0.930672 (0.000475) with: {'learning_rate': 0.01, 'n_estimators': 50, 'subsample': 0.5}
0.913739 (0.090129) with: {'learning_rate': 0.01, 'n_estimators': 50, 'subsample': 0.9}
0.955637 (0.003901) with: {'learning_rate': 0.01, 'n_estimators': 100, 'subsample': 0.5}
0.955391 (0.004388) with: {'learning_rate': 0.01, 'n_estimators': 100, 'subsample': 0.9}
0.969867 (0.004868) with: {'learning_rate': 0.01, 'n_estimators': 500, 'subsample': 0.5}
0.968961 (0.004865) with: {'learning_rate': 0.01, 'n_estimators': 500, 'subsample': 0.9}
0.956156 (0.003892) with: {'learning_rate': 0.1, 'n_estimators': 10, 'subsample': 0.5}
0.956554 (0.003652) with: {'learning_rate': 0.1,

In [56]:
grid_result

GridSearchCV(cv=RepeatedStratifiedKFold(n_repeats=3, n_splits=10, random_state=1),
             error_score=nan,
             estimator=GradientBoostingClassifier(ccp_alpha=0.0,
                                                  criterion='friedman_mse',
                                                  init=None, learning_rate=0.1,
                                                  loss='deviance', max_depth=3,
                                                  max_features=None,
                                                  max_leaf_nodes=None,
                                                  min_impurity_decrease=0.0,
                                                  min_impurity_split=None,
                                                  min_samples_leaf=1,
                                                  min_samples_spl...
                                                  n_iter_no_change=None,
                                                  presort='deprecated',
          

In [ ]:
grid = dict()
grid['n_estimators'] = [100, 500, 750, 1000]
grid['learning_rate'] = [0.001, 1]
grid['subsample'] = [0.5, 0.9]
scores = ["precision", "recall"]
# define the model with default hyperparameters
model = GradientBoostingClassifier()
# define the evaluation procedure
cv = RepeatedStratifiedKFold(n_splits=10, n_repeats=3, random_state=1)

# define the grid search procedure
for score in scores:
    print("# Tuning hyper-parameters for %s" % score)
    print()
    grid_search = GridSearchCV(estimator=model, param_grid=grid, n_jobs=-1, cv=cv, scoring="%s_macro" % score)
    # execute the grid search
    grid_result = grid_search.fit(X_tr, y_tr)
    # summarize the best score and configuration
    print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
    # summarize all scores that were evaluated
    print("Grid scores on development set:")
    print()
    means = grid_result.cv_results_['mean_test_score']
    stds = grid_result.cv_results_['std_test_score']
    params = grid_result.cv_results_['params']
    for mean, stdev, param in zip(means, stds, params):
      print("%f (%f) with: %r" % (mean, stdev, param))
    print()

    print("Detailed classification report:")
    print()
    print("The model is trained on the full development set.")
    print("The scores are computed on the full evaluation set.")
    print()
    y_true, y_pred = y_te, grid_search.predict(X_te)
    print(classification_report(y_true, y_pred))
    print()

# Tuning hyper-parameters for precision

Best: 0.955917 using {'learning_rate': 0.001, 'n_estimators': 1000, 'subsample': 0.5}
Grid scores on development set:

0.428521 (0.000098) with: {'learning_rate': 0.001, 'n_estimators': 100, 'subsample': 0.5}
0.428521 (0.000098) with: {'learning_rate': 0.001, 'n_estimators': 100, 'subsample': 0.9}
0.930625 (0.000485) with: {'learning_rate': 0.001, 'n_estimators': 500, 'subsample': 0.5}
0.880359 (0.150655) with: {'learning_rate': 0.001, 'n_estimators': 500, 'subsample': 0.9}
0.954637 (0.003324) with: {'learning_rate': 0.001, 'n_estimators': 750, 'subsample': 0.5}
0.954760 (0.003286) with: {'learning_rate': 0.001, 'n_estimators': 750, 'subsample': 0.9}
0.955917 (0.003495) with: {'learning_rate': 0.001, 'n_estimators': 1000, 'subsample': 0.5}
0.955453 (0.004372) with: {'learning_rate': 0.001, 'n_estimators': 1000, 'subsample': 0.9}
0.884076 (0.090072) with: {'learning_rate': 1, 'n_estimators': 100, 'subsample': 0.5}
0.902604 (0.081096) with: {'lear